internal/pkg/meta/meta.go

```diff
@@ -0,0 +1,317 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package meta provides access to META partition: key-value partition persisted across reboots.
+package meta
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-blockdevice/blockdevice/probe"
+
+	"github.com/siderolabs/talos/internal/pkg/meta/internal/adv"
+	"github.com/siderolabs/talos/internal/pkg/meta/internal/adv/syslinux"
+	"github.com/siderolabs/talos/internal/pkg/meta/internal/adv/talos"
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+	"github.com/siderolabs/talos/pkg/machinery/resources/runtime"
+)
+
+// Meta represents the META reader/writer.
+//
+// Meta abstracts away all details about loading/storing the metadata providing an easy to use interface.
+type Meta struct {
+	legacy adv.ADV
+	talos  adv.ADV
+	state  state.State
+	opts   Options
+}
+
+// Options configures the META.
+type Options struct {
+	fixedPath string
+}
+
+// Option is a functional option.
+type Option func(*Options)
+
+// WithFixedPath sets the fixed path to META partition.
+func WithFixedPath(path string) Option {
+	return func(o *Options) {
+		o.fixedPath = path
+	}
+}
+
+// New initializes empty META, trying to probe the existing META first.
+func New(ctx context.Context, st state.State, opts ...Option) (*Meta, error) {
+	meta := &Meta{
+		state: st,
+	}
+
+	for _, opt := range opts {
+		opt(&meta.opts)
+	}
+
+	var err error
+
+	meta.legacy, err = syslinux.NewADV(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	meta.talos, err = talos.NewADV(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	err = meta.Reload(ctx)
+	if err != nil && !os.IsNotExist(err) {
+		return meta, err
+	}
+
+	return meta, nil
+}
+
+func (meta *Meta) getPath() (string, error) {
+	if meta.opts.fixedPath != "" {
+		return meta.opts.fixedPath, nil
+	}
+
+	dev, err := probe.GetDevWithPartitionName(constants.MetaPartitionLabel)
+	if err != nil {
+		return "", err
+	}
+
+	defer dev.Close() //nolint:errcheck
+
+	return dev.PartPath(constants.MetaPartitionLabel)
+}
+
+// Reload refreshes the META from the disk.
+func (meta *Meta) Reload(ctx context.Context) error {
+	path, err := meta.getPath()
```

Comment:
do we need locking here, if some controller uses this (haven;t looked at the whole changes yet, just thinking if controllers use this?)

Reply:
controllers do not use it... but good point on locking... probably to be on the safe side I would rather use locks

Reply:
added locks